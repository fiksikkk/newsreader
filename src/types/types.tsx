import { NativeStackScreenProps } from '@react-navigation/native-stack';

export type StackParamList = {
  Home: undefined;
  Character: { userid: number | string };
  CreateNewCharacter: { createCharacter: boolean};
};

export type HomeProps = NativeStackScreenProps<StackParamList, 'Home'>;
export type CharacterProps = NativeStackScreenProps<StackParamList,'Character'>;
export type CreacteCharacterProps = NativeStackScreenProps<StackParamList, 'CreateNewCharacter'>;

export interface Character {
  name: string;
  image: string;
  id: number;
}

interface Info {
  next: number;
  prev: number;
  pages: number;
  count: number;
}

export interface Characters {
  characters: { results: Character[]; info: Info };
}

export interface GetPostsVariables {
  page: number
}