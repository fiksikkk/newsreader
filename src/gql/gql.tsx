import {gql} from '@apollo/client';

export const GET_POSTS = gql`
  query {
    characters {
      info {
        count
      }
      results {
        id
        name
        image
      }
    }
  }
`;

export const GET_CHARACTER = (id: number | string) => {
  return gql`
    query {
      character(id: ${id}) {
        name
        status
        species
        type
        gender
        image
      }      
    }
  `;
};
