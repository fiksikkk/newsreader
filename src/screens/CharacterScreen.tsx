import React from 'react';

import {useQuery} from '@apollo/client';
import {Image, StyleSheet, Text, View} from 'react-native';

import {GET_CHARACTER} from '../gql/gql';
import {CharacterProps} from '../types/types';

const CharacterScreen = ({route}: CharacterProps) => {
  const {loading, error, data} = useQuery(GET_CHARACTER(route.params.userid));

  if (loading) {
    return <Text testID='progress'>Loading...</Text>;
  }
  if (error) {
    return <Text testID='error'>Error : {error.message}</Text>;
  }

  return (
    <View testID='container' style={styles.container} key={data.character.id}>
      <Image style={styles.image} src={data.character.image} />
      <Text style={styles.name}>Name: {data.character.name}</Text>
      <Text style={styles.name}>Gender: {data.character.gender}</Text>
      <Text style={styles.name}>Status: {data.character.status}</Text>
      <Text style={styles.name}>Species: {data.character.species}</Text>
      <Text style={styles.name}>Type: {data.character.type}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  image: {
    width: 350,
    height: 350,
  },
  container: {
    padding: 25,
    margin: 10,
    flexDirection: 'column',
  },
  name: {
    fontSize: 24,
    color: 'black',
  },
});

export default CharacterScreen;
