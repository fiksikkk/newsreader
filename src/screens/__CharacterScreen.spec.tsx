import React from 'react';

import { MockedProvider } from "@apollo/client/testing";
import { render, waitFor } from "@testing-library/react-native";
import { GET_CHARACTER, GET_ERROR, GET_POSTS } from "../gql/gql";
import CharacterScreen from "./CharacterScreen";
import { CharacterProps } from '../types/types';

describe('Should render CharacterScreen', () => {
  const navigation = { navigate: jest.fn() } as unknown as CharacterProps['navigation'];
  const route = { route: jest.fn(), params: { userid: 7 } } as unknown as CharacterProps['route'];


  it('Shold render and show progress on loading', () => {
    const utils = render(
      <MockedProvider addTypename={false} mocks={[mock(7)]}>
        <CharacterScreen navigation={navigation} route={route} />
      </MockedProvider>
    );

    expect(utils.queryByTestId('progress')).toBeTruthy();
  });

  it('shold render and show an error', async () => {
    const utils = render(
      <MockedProvider addTypename={false} mocks={[errorMock(7)]}>
        <CharacterScreen navigation={navigation} route={route} />
      </MockedProvider>
    );
    await waitFor(() => [
      expect(utils.queryByTestId('progress')).toBeFalsy(),
    ]);
    expect(utils.queryByTestId('error')).toBeTruthy();
  })

  it('Should render a cards when loading is done', async () => {
    const utils = render(
      <MockedProvider addTypename={false} mocks={[mock(7)]}>
        <CharacterScreen navigation={navigation} route={route} />
      </MockedProvider>
    );

    await waitFor(() => [
      expect(utils.queryByTestId('progress')).toBeFalsy()
    ]);

    expect(utils.queryByTestId('container')?.children.length).toBe(6);

    expect(utils.toJSON()).toMatchSnapshot();
  });
});

const mock = (id: number) => {
  return {
    request: {
      query: GET_CHARACTER, variables: { id }
    },
    result: {
      data: {
        "character": {
          "name": "Abradolf Lincler",
          "status": "unknown",
          "species": "Human",
          "type": "Genetic experiment",
          "gender": "Male",
          "image": "https://rickandmortyapi.com/api/character/avatar/7.jpeg"

        }
      }
    }
  }
}

const errorMock = (id: number) => {
  return {
    request: {
      query: GET_CHARACTER, variables: { id }
    },
    result: {
      errors: [
        {
          message: "Syntax Error: Expected Name, found <EOF>",
          locations: [
            {
              line: 4,
              column: 6
            }
          ],
          extensions: {
            code: "GRAPHQL_PARSE_FAILED"
          }
        }
      ]
    }
  }
}