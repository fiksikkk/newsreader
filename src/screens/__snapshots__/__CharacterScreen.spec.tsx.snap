// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render CharacterScreen Should render a cards when loading is done 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "margin": 10,
      "padding": 25,
    }
  }
  testID="container"
>
  <Image
    src="https://rickandmortyapi.com/api/character/avatar/7.jpeg"
    style={
      {
        "height": 350,
        "width": 350,
      }
    }
  />
  <Text
    style={
      {
        "color": "black",
        "fontSize": 24,
      }
    }
  >
    Name: 
    Abradolf Lincler
  </Text>
  <Text
    style={
      {
        "color": "black",
        "fontSize": 24,
      }
    }
  >
    Gender: 
    Male
  </Text>
  <Text
    style={
      {
        "color": "black",
        "fontSize": 24,
      }
    }
  >
    Status: 
    unknown
  </Text>
  <Text
    style={
      {
        "color": "black",
        "fontSize": 24,
      }
    }
  >
    Species: 
    Human
  </Text>
  <Text
    style={
      {
        "color": "black",
        "fontSize": 24,
      }
    }
  >
    Type: 
    Genetic experiment
  </Text>
</View>
`;
